mmvec/ALR.py

```diff
@@ -0,0 +1,173 @@
+import pandas as pd
+
+import torch
+from torch import linalg
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.distributions import Multinomial, Normal
+
+from skbio import OrdinationResults
+
+
+def structure_data(microbes, metabolites):
+    #microbes = microbes.to_dataframe().T
+    #metabolites = metabolites.to_dataframe().T
+    microbes = microbes.loc[metabolites.index]
+
+    microbe_idx = microbes.columns
+    metabolite_idx = metabolites.columns
+
+    microbe_count = microbes.shape[1]
+    metabolite_count = metabolites.shape[1]
+
+    microbes = torch.tensor(microbes.values, dtype=torch.int)
+    metabolites = torch.tensor(metabolites.values, dtype=torch.int64)
+
+    microbe_relative_frequency = (microbes.T/microbes.sum(1)).T
+
+    return (microbes, metabolites, microbe_idx, metabolite_idx, microbe_count,
+           metabolite_count, microbe_relative_frequency)
+
+
+class LinearALR(nn.Module):
+    def __init__(self, input_dim, output_dim):
+        super().__init__()
+        self.linear = nn.Linear(input_dim, output_dim - 1)
+
+    def forward(self, x):
+        y = self.linear(x)
+        z = torch.zeros((y.shape[0], y.shape[1], 1))
+        y = torch.cat((z, y), dim=2)
+
+        return F.softmax(y, dim=2)
+
+
+class MMvecALR(nn.Module):
+    def __init__(self, microbes, metabolites, latent_dim, sigma_u=1,
+                 sigma_v=1):
+        super().__init__()
+
+        # Data setup
+        (self.microbes, self.metabolites,
+         self.microbe_idx, self. metabolite_idx,
+         self.num_microbes, self.num_metabolites,
+         self.microbe_relative_freq) = structure_data(microbes,
+                metabolites)
+        self.sigma_u = sigma_u
+        self.sigma_v = sigma_v
+        self.latent_dim = latent_dim
+        # TODO: intialize same way as linear bias
+        self.encoder_bias = nn.parameter.Parameter(
+                torch.randn((self.num_microbes, 1)))
+
+        self.encoder = nn.Embedding(self.num_microbes, self.latent_dim)
+        self.decoder = LinearALR(self.latent_dim, self.num_metabolites)
+
+
+    def forward(self, X):
+        # Three likelihoods, the likelihood of each weight and the likelihood
+        # of the data fitting in the way that we thought
+        # LYs
+        z = self.encoder(X)
+        z = z + self.encoder_bias[X].reshape((*X.shape, 1))
+        y_pred = self.decoder(z)
+
+        forward_dist = Multinomial(total_count=0,
```

Comment:
I'd suggest getting rid of the `total_count=0` parameter -- we don't actually need it for `log_prob`.
And it may introduce a bug downstream (since the total_count isn't actually zero).

Reply:
This was actually a result of doing things in batch. We would run into an issue where the log_prob would indicate our calculation was out of the support of the distribution, because it had different counts sample to sample, so we solved it via this suggestion:

https://github.com/pytorch/pytorch/issues/42407#issuecomment-668089726

That said, looking at the documentation again, I wonder if we should be using `logits` instead of `probs`?